src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/controller/Cas2ApplicationsController.kt

```diff
@@ -0,0 +1,159 @@
+      return ResponseEntity.ok(getPersonDetailAndTransform(application, user))
+    }
+    throw NotFoundProblem(applicationId, "Application")
+  }
+
+  @Transactional
+  override fun cas2ApplicationsPost(body: NewApplication, createWithRisks: Boolean?):
+    ResponseEntity<Application> {
+    val deliusPrincipal = httpAuthService.getDeliusPrincipalOrThrow()
+    val user = userService.getUserForRequest()
+
+    val personInfo = offenderService.getFullInfoForPersonOrThrow(body.crn, user)
+
+    val applicationResult = applicationService.createCas2Application(body.crn, user, deliusPrincipal.token.tokenValue)
+
+    val application = when (applicationResult) {
+      is ValidatableActionResult.GeneralValidationError -> throw BadRequestProblem(errorDetail = applicationResult.message)
+      is ValidatableActionResult.FieldValidationError -> throw BadRequestProblem(invalidParams = applicationResult.validationMessages)
+      is ValidatableActionResult.ConflictError -> throw ConflictProblem(id = applicationResult.conflictingEntityId, conflictReason = applicationResult.message)
+      is ValidatableActionResult.Success -> applicationResult.entity
+    }
+
+    return ResponseEntity
+      .created(URI.create("/applications/${application.id}"))
+      .body(applicationsTransformer.transformJpaToApi(application, personInfo))
+  }
+
+  @Transactional
+  override fun cas2ApplicationsApplicationIdPut(
+    applicationId: UUID,
+    body:
+      UpdateApplication,
+  ): ResponseEntity<Application> {
+    val user = userService.getUserForRequest()
+
+    val serializedData = objectMapper.writeValueAsString(body.data)
+
+    val applicationResult = applicationService.updateCas2Application(applicationId = applicationId, data = serializedData, username = user.deliusUsername)
+
+    val validationResult = when (applicationResult) {
+      is AuthorisableActionResult.NotFound -> throw NotFoundProblem(applicationId, "Application")
+      is AuthorisableActionResult.Unauthorised -> throw ForbiddenProblem()
+      is AuthorisableActionResult.Success -> applicationResult.entity
+    }
+
+    val updatedApplication = when (validationResult) {
+      is ValidatableActionResult.GeneralValidationError -> throw BadRequestProblem(errorDetail = validationResult.message)
+      is ValidatableActionResult.FieldValidationError -> throw BadRequestProblem(invalidParams = validationResult.validationMessages)
+      is ValidatableActionResult.ConflictError -> throw ConflictProblem(id = validationResult.conflictingEntityId, conflictReason = validationResult.message)
+      is ValidatableActionResult.Success -> validationResult.entity
+    }
+
+    return ResponseEntity.ok(getPersonDetailAndTransform(updatedApplication, user))
+  }
+
+  override fun cas2ApplicationsApplicationIdSubmissionPost(
+    applicationId: UUID,
+    submitApplication: SubmitCas2Application,
+  ): ResponseEntity<Unit> {
+    val deliusPrincipal = httpAuthService.getDeliusPrincipalOrThrow()
```

Comment:
```suggestion
    httpAuthService.getDeliusPrincipalOrThrow()
```
do we need to get the username/`deliusPrincipal` as variables here?  
(I guess we don't really need to call `getDeliusPrincipalOrThrow` for CAS2, but we might want to leave it in so we can replace with something that checks we're authorising with NOMIS later on?)

Reply:
Good point, we certainly aren't using that `deliusPrincipal` or its `username` but the `getDeliusPrincipalOrThrow()` is an important control in the current implementation.

I'll replace:

```kotlin
val deliusPrincipal = httpAuthService.getDeliusPrincipalOrThrow()
val username = deliusPrincipal.name
```
with

```kotlin
val _deliusPrincipal = httpAuthService.getDeliusPrincipalOrThrow()
```

which as you suggest might be a useful reminder...